proposals/NNNN-bitfield-insert-extract.md

```diff
@@ -0,0 +1,116 @@
+given logical shifting / masking operation as a bitfield extraction / insertion 
+routine. ([afaik, this never happens](https://github.com/microsoft/DirectXShaderCompiler/issues/6902)). As a result, the `Lbfe`, `Ubfe`, 
+and `Bfi` instructions currently go unused. 
+
+Another application where bitfield extraction and insertion is useful is in
+efficient "reinterpretation" of structures. (For example, accessing the fourth
+byte of a 32-bit integer as if it were an int8_t4). We would like to use these
+intrinsics within [Slang](https://github.com/shader-slang/slang/issues/4817) to 
+help with casting from one structure layout to another, where a mix of 8, 16, 
+32, and 64-bit type sizes might be used. 
+
+Beyond this, certain priority queues and stacks depend on the "sign extension"
+behavior that bitfield intrinsics expose, which enable reducing register pressure
+for N-wide data structure traversals. This can be very helpful for compute shader
+operations, particularly for two-wide BVH to N-wide BVH conversions like what's 
+done in [H-PLOC](https://gpuopen.com/download/publications/HPLOC.pdf), or like 
+the N-wide stack entries proposed by Ylitie.
+
+In short, HLSL is missing bitfield insertion and extraction intrinsics, and so 
+at the moment there doesn't seem to be a way to leverage the corresponding DXIL
+intrinsics. 
+
+## Proposed solution
+
+This proposal would introduce `bitfieldExtract` and `bitfieldInsert` intrinsics, 
+which would map to their corresponding DXIL instructions. For HLSL->SPIR-V compilation, 
+we would directly map these intrinsics to the corresponding SPIR-V operations
+(namely, `OpBitfieldSExtract`, `OpBitfieldUExtract`, and `OpBitfieldInsert`).
+
+In the case of bitfield extraction, the logical expression:
+```
+int value = ..., offset = ..., bits = ...;
+int extracted = (((value>>offset)&((1u<<bits)-1))<<(32-bits)>>(32-bits));
+```
+Would then simplify to
+```
+int value = ..., offset = ..., bits = ...;
+int extracted = bitfieldExtract(value, offset, bits);
+```
+Where the resulting instruction would involve `Lbfe`.
+
+If value is unsigned, then the logical expression:
+```
+uint value = ..., offset = ..., bits = ...;
+uint extracted = ((value>>offset)&((1u<<bits)-1));
+```
+Would then simplify to
+```
+uint value = ..., offset = ..., bits = ...;
+int extracted = bitfieldExtract(value, offset, bits);
+```
+Where the resulting instruction would involve `Ubfe`.
+
+Finally, in the case of bitfield insertion, the logical expression:
+```
+uint clearMask = ~(((1u << bits) - 1u) << offset);
+uint clearedBase = base & clearMask;
+uint maskedInsert = (insert & ((1u << bits) - 1u)) << offset;
+uint result = clearedBase | maskedInsert; 
+```
+Would then simplify to
+```
+uint base = ..., insert = ..., offset = ..., bits = ...;
+int extracted = bitfieldInsert(base, insert, value, bits);
+```
+Where the resulting instruction would involve `Bfi`.
+
+The `bitfieldExtract` intrinsic would support a component-wise value, for 
```

Comment:
Will matrix type like uint2x2 be supported?

Reply:
I believe the intrinsics only work component-wise for vector types. For matrix types, I could easily enough use the intrinsic over the rows. We’d probably want to limit the scope of this to what GLSL can do, and not much beyond this.

Reply:
Ah, right, so in GLSL, there are no integer matrix types. So then for a type like `uint2x2`, the expected usage would be something like:

```
uint2x2 matrix = uint2x2(...);
uint4 temp = bitfieldExtract(uint4(matrix[0], matrix[1]), offset, numBits);
uint2x2 bfeMatrix = uint2x2(temp.xy, temp.zw);
```

With some syntax sugar, it would theoretically be possible to do something like:
```
uint2x2 matrix = uint2x2(...);
uint2x2 bfeMatrix = bitfieldExtract(matrix, offset, numBits);
```
But iiuc a swizzle operation would be needed, which would be more expensive performance wise (though not by much) rather than directly using a uint4.  

Since this is more about enabling low level "ninja optimizations" in specialized codes, I'd lean towards keeping the feature trimmed down to just scalars and vectors, unless you have a particular application in mind.